### PR TITLE
Adding a TreeScan Gate Removal pass to parallelize Scanning Gate

### PR DESCRIPTION
**Pull request by jkalloor3**
For several circuits of medium width 7-20ish qubits, GPU-enhanced QFactor allows us to run instantiation quickly in order to do a large-block ScanningGate removal. For these workflows, there are very few blocks, which limits our parallelism. In order to fix this, this PR adds a tree structure to ScanningGate removal. The TreeScan works as follows:

1. Split the circuit operations into chunks of size tree_depth
2. At every iteration:
a. Look at the next chunk of operations 
b. Generate 2 ^ tree_depth circuits. Each circuit corresponds to every combination of whether or not to include one of the operations in the chunk.
c. Instantiate in parallel all 2^tree_depth circuits
d. Choose the circuit that has the least number of operations and move on to the next chunk of operations.

This optimization is less greedy than the current ScanningGate removal, which we see can offer much better quality circuits than ScanningGate. In very rare occasions, ScanningGate may be able to outperform TreeScan (since it is still greedy), but in general we can expect TreeScan to almost always outperform ScanningGate.

### CODE REVIEW

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:24]: Update documentation, most of your PR comment can fit well here

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:59]: What is the default? Can you give a more user-friendly example of what a specific number means? Like what if I set this to three, what does it do there? If this is adequately described in the class docstring, then you can go less verbose here. Also should the default be 1, isn't that just scanning?

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:95]: Add type check. You can use `bqskit.utils.typing.is_integer`

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:109]: To keep with the rest of the repo's style, we do one parameter per line in multi-line parameters. No need to change, but worth mentioning

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:101]: `min_iters` can probably be 10 or even 0

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:136]: Update log msg

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:139]: Remove unnecessary line

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:147]: The second string in the multiline string is not an f-string. You need to prepend `f` on all strings in the multiline.

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:154]: "Circuits of size" is a bit misleading, it might be better to say "2 ** tree_depth - 1 Circuits".

Moreover, this comment would be redundant with a proper docstring for `get_tree_circs`

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:160]: This should be fused into the function. You can create a wrapper function that executes the two lines of code above. Not that important. Your call to do it or not...

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:164]: I'm not sure if this is descriptive of what is occurring or proper grammar. It might be better to say, "Attempting removal of up to X operations."

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:181]: From a logging perspective, this is not helpful. We should instead say something like... "Successfully removed X CNOTs Y U3s" or something like that. Plus f-string issue.

Consider a situation where a user is running on Perlmutter for 12 hours and the time-limit hits. You go to the logs to try and gather some amount of information about how it went to see if its worth restarting or if it did well. You look at the logs and you see a bunch of "Successfully switched to circuit X of of {2 ** Y}." That may not be helpful. Instead if I can count how many successfully removed CNOTs, then I can gather that in 12 hours I made Z amount of progress or something. I know this means that you will have to do more than a simple one-line log call to gather this information, but it is worth doing good logging.

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:111]: - Remove comment, also this isn't recursive
- Add docstring

File: bqskit/passes/__init__.py
Comment (edyounis) [inline bqskit/passes/__init__.py:333]: Also add the to the documentation docstring in this file

**Review by edyounis — CHANGES_REQUESTED** [review]
Good PR. Minor changes requested

File: bqskit/passes/__init__.py
Comment (jkalloor3) [reply bqskit/passes/__init__.py:334]: Does that just mean adding the TreeScanningGateRemovalPass where I did in my latest commit?

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:1]: Update Docstring

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:23]: Update Docstring

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:27]: When writing documentation, it's good practice to wrap variable names in backticks. Like tree_depth should be `tree_depth`

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:36]: For better documentation, you may want to link ScanningGateRemoval using the proper syntax:

```raw
:class:`~bqskit.passes.processing.ScanningGateRemovalPass`
```
We can also double check the documentation before merging by clicking on the PR readthedocs test.

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:37]: offer much

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:38]: I don't understand why ScanningGate would be able to outperform TreeScan. Isn't every candidate considered during ScanningGate considered during TreeScan? I know there is always variance in instantiation so that anything can happen for sure, but is there something I am missing?

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:53]: Update Docstring

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [inline bqskit/passes/processing/treescan.py:140]: Function and method docstrings should start with a single sentence summary, then more if needed after a blank line, followed by argument descriptions. More sections can follow if necessary, e.g., Returns, Examples, See Also, etc. I know this is just a helper method now, but I can foresee this method being reused in future projects. That is why I am being pedantic here. We should make it as readable as possible.

File: bqskit/passes/processing/treescan.py
Comment (edyounis) [reply bqskit/passes/processing/treescan.py:165]: Log should indicate this is a tree scan pass.

**Review by edyounis — CHANGES_REQUESTED** [review]
You did not address a few comments from the previous review. I am not a fan of how GitHub collapses the comments, so you probably missed them. You might need to expand the ones in the middle. Let me know if you need help here.

Otherwise, it looks much better. All of my comments this time around were focused on the documentation. We should make sure the code you write is properly documented; it will ensure others can use, maintain, and expand your code. Once you push up a commit, you can check how the documentation looks by clicking on the readthedocs CI check.

File: bqskit/passes/processing/treescan.py
Comment (jkalloor3) [reply bqskit/passes/processing/treescan.py:38]: For example, if there are 20 gates and TreeScan looks at 5 at a time. Let's say that the options are that if you remove the first gate, you can also remove the last 10 gates in the circuit. Otherwise, if you don't remove the first gate, you can remove the next 4 gates in the circuit (but this removes your ability to remove the last 10 gates in the circuit). TreeScan intially would pick the 4 gate option over the single gate option, and then miss out removing the last 10 gates. I don't know if this is even possible in practice, and I haven't seen a single case where this does happen, but AFAIK it could be a possibility?

Comment (edyounis) [reply bqskit/passes/processing/treescan.py:39]: I see. Yeah, that makes sense. Thanks for explaining.

**Pull request merged**
